utils/graph_optimize.py

```diff
@@ -97,7 +97,109 @@ def merge_bn(symbol, args, auxs, symbol_only=False):
     outputs = outputs[0] if len(outputs) == 1 else mx.sym.Group(outputs)
     return outputs, args, auxs
 
+def attach_quantize_node(symbol, out_shape_dict, base_quant_attrs, quantized_op=["Convolution", "FullyConnected", "Deconvolution"]):
+    """
+    Adapted from https://github.com/dmlc/tvm/blob/master/python/tvm/relay/frontend/mxnet.py
+    Instead of translating nnvm graph into TVM relay graph, we adapt the script to translate
+    it back to mxnet graph.
+    """
+    assert symbol is not None
+    assert base_quant_attrs is not None
+    # currently only support quant_mode = "minmax" and weight per tensor quantization method
+    base_quant_attrs["is_weight"] = "False"
+    base_quant_attrs["is_weight_perchannel"] = "False"
+    base_quant_attrs["quant_mode"] = "minmax"
+
+    data_quant_attrs = base_quant_attrs.copy()
+    weight_quant_attrs = base_quant_attrs.copy()
+    weight_quant_attrs["is_weight"] = "True"
+
+    jgraph = json.loads(symbol.tojson())
+    jnodes = jgraph["nodes"]
+    node_map = {}
+    node_op_map = {}
+    quantized_node_map = {}
+
+
+    for nid, node in enumerate(jnodes):
+        # edges are [which_node, which_output, type(? not sure)]
+        # mx.symbol has an attribute of __getitem__. sym[1] gives the second output
+        children = [node_map[e[0]][e[1]] for e in node["inputs"]]
+        attrs = node.get("attrs", {})
+        node_name = node["name"]
+        op_name = node["op"]
+        if op_name == "null":
+            attrs = dict({k:v for k, v in attrs.items() if k.startswith("__")})
+            assert node_name in out_shape_dict.keys(), "{} Variable is not in shape_dict".format(node_name)
+            if "__shape__" not in attrs.keys():
+                attrs["__shape__"] = out_shape_dict[node_name]
+                attrs["__dtype__"] = 0  # "float32"
```

Comment:
Could this 0 be changed to something more informative? Like defining a constatnt FLOAT32_DTYPE = 0.